src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -594,6 +594,21 @@ static uint getDeclarationValEscape(BoundTypeExpression typeExpression, uint val
         public override BoundNode? VisitRecursivePattern(BoundRecursivePattern node)
         {
             SetPatternLocalScopes(node);
+
+            // If the equivalent `Deconstruct` call has unscoped receiver,
+            // we narrow the escape scope of the pattern input
+            // which flows as the escape scope of all ref-struct declaration subpatterns
+            // and so the ref safety of the pattern is equivalent to a `Deconstruct(out var ...)` invocation
+            // where "safe-context inference of declaration expressions" would have the same effect.
+            if (node.DeconstructMethod is { HasUnscopedRefAttribute: true } or
+                { IsExtensionMethod: true, Parameters: [{ EffectiveScope: ScopedKind.None }, ..] })
+            {
+                using (new PatternInput(this, _localScopeDepth))
+                {
+                    return base.VisitRecursivePattern(node);
+                }
+            }
```

Comment:
Looking through the spec some more, makes me wonder: what if the `Deconstruct` method was compiled with an older compiler? Then the receiver, which is treated as a `ref` parameter, would be considered unscoped by default.

Here's the relevant [section](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/low-level-struct-improvements.md#refsafetyrulesattribute):
> To reduce the chance of breaking changes when recompiling with C#11, we will update the C#11 compiler to use the ref safe context rules for method invocation that match the rules that were used to analyze the method declaration. Essentially, when analyzing a call to a method compiled with an older compiler, the C#11 compiler will use C#7.2 ref safe context rules.

Reply:
> Let me try to summarize my understanding. Does that match your understanding?

Yes, your summary matches my understanding.

> what if the `Deconstruct` method was compiled with an older compiler? Then the receiver, which is treated as a `ref` parameter, would be considered unscoped by default.

Good point. I think if I rewrite the condition in terms of EffectiveScope as you suggested in the other comment, and we take https://github.com/dotnet/roslyn/pull/75930, this scenario will work (because that PR makes EffectiveScope consider old/updated rules).

Reply:
Using `EffectiveScope` seems a good solution to abstract away from those differences.
Were you able to add a test for this old rules vs. new rules scenario?

Reply:
I'm waiting for the other PR to be merged before adding a test.

Reply:
> what if the `Deconstruct` method was compiled with an older compiler? Then the receiver, which is treated as a `ref` parameter, would be considered unscoped by default.

Actually this does not seem to be true. It's hard to tell from the spec how it should be because the old rules have no concept of scoped/unscoped. Anyway, I added tests that verify the behavior is consistent between explicit calls and patterns, so it behaves as expected wrt this PR, I think.